### ISSUE TITLE
%f{depth} filename separator

### ISSUE THREAD

**Issue by CarlosGomez-dev** [log4js-node/log4js-node#1060]
Hello, it seems I can't get `%f{depth}` to work on my system (Windows 10, Node v14.15), here's a sample codeblock:

```javascript
import log4js from 'log4js';

log4js.configure({
  appenders: {
    out: {
      type: 'console',
      layout: {
        type: 'pattern',
        pattern: '%d %f{1} %m',
      },
    },
  },
  categories: {
    default: { appenders: ['out'], level: 'info', enableCallStack: true },
  },
});

const logger = log4js.getLogger('app');

logger.info('Test log message');
```
Expected output:
`2021-02-22T12:46:30.039 /main.js Test log message`
Actual output:
`2021-02-22T12:46:30.039 file:///C:/Users/Username/Desktop/Projects/Logtest/src/main.js Test log message`

The issue seems to be in `layouts.js` because `path.sep` is ([correctly](https://nodejs.org/api/path.html#path_path_sep)) returning `\`, while filename is `file:///C:/Users/Username/Desktop/Projects/Logtest/src/main.js`, so the filename is never split.

Is this an OS and/or node misconfiguration on my part?

I managed to correct the issue by modifying `layouts.js` split separator parameter, I'll submit a fix PR in case this is a general issue.

**Comment by peteriman** [comment]
I ran your exact codes (saved into test.js) using log4js 6.3.0 on Windows 10 with Node.js 14.
## `node test.js`
> v14.15.0 - `2022-01-04T21:45:59.119 test.js Test log message`
> v14.15.1 - `2022-01-04T21:46:05.225 test.js Test log message`
> v14.15.2 - `2022-01-04T21:46:08.476 test.js Test log message`
> v14.15.3 - `2022-01-04T21:46:13.812 test.js Test log message`
> v14.15.4 - `2022-01-04T21:46:15.480 test.js Test log message`
> v14.15.5 - `2022-01-04T21:46:18.274 test.js Test log message`


Doesn't seem to have any issues.

----

I edited the [layouts.js](https://github.com/log4js-node/log4js-node/blob/master/lib/layouts.js) to have some debug logs:
```javascript
  function fileName(loggingEvent, specifier) {
    let filename = loggingEvent.fileName || '';
    console.log("debug-full-filename:", filename);             // added
    console.log("debug-path.sap:", path.sep);                  // added
    if (specifier) {
      const fileDepth = parseInt(specifier, 10);
      const fileList = filename.split(path.sep);
      if (fileList.length > fileDepth) {
        filename = fileList.slice(-fileDepth).join(path.sep);
        console.log("debug-trunc-filename:", filename);        // added
      }
    }

    return filename;
  }
```

## `node test.js`

> `debug-full-filename: C:\Users\user\Desktop\log4js-node\test.js`
> `debug-path.sap: \`
> `debug-trunc-filename: test.js`
> `2022-01-04T21:52:38.144 test.js Test log message`
---

I noticed a difference between yours and mine:
`file:///C:/Users/Username/Desktop/Projects/Logtest/src/main.js` vs `C:\Users\user\Desktop\log4js-node\test.js`

**Comment by peteriman** [comment]
I ran your codes, with some modifications saved into `test.js`, using `log4js 6.3.0` on `Windows 10` with `Node.js 14`.

```javascript
const log4js = require('./lib/log4js'); // changed

log4js.configure({
  appenders: {
    out: {
      type: 'console',
      layout: {
        type: 'pattern',
        pattern: '%d %f{1} %m',
      },
    },
  },
  categories: {
    default: { appenders: ['out'], level: 'info', enableCallStack: true },
  },
});

const logger = log4js.getLogger('app');

logger.info('Test log message', __filename); // changed
```


## `node test.js`
> v14.15.0 - `2022-01-04T21:45:59.119 test.js Test log message C:\Users\user\Desktop\log4js-node\test.js`
> v14.15.1 - `2022-01-04T21:46:05.225 test.js Test log message C:\Users\user\Desktop\log4js-node\test.js`
> v14.15.2 - `2022-01-04T21:46:08.476 test.js Test log message C:\Users\user\Desktop\log4js-node\test.js`
> v14.15.3 - `2022-01-04T21:46:13.812 test.js Test log message C:\Users\user\Desktop\log4js-node\test.js`
> v14.15.4 - `2022-01-04T21:46:15.480 test.js Test log message C:\Users\user\Desktop\log4js-node\test.js`
> v14.15.5 - `2022-01-04T21:46:18.274 test.js Test log message C:\Users\user\Desktop\log4js-node\test.js`

Doesn't seem to have any issues.

----

I edited the [lib/layouts.js](https://github.com/log4js-node/log4js-node/blob/master/lib/layouts.js) to have some debug logs:
```javascript
  function fileName(loggingEvent, specifier) {
    let filename = loggingEvent.fileName || '';
    console.log("debug-full-filename:", filename);            // added
    console.log("debug-path.sep:", path.sep);                 // added
    if (specifier) {
      const fileDepth = parseInt(specifier, 10);
      const fileList = filename.split(path.sep);
      if (fileList.length > fileDepth) {
        filename = fileList.slice(-fileDepth).join(path.sep);
        console.log("debug-trunc-filename:", filename);       // added
      }
    }

    return filename;
  }
```

## `node test.js`

> `debug-full-filename: C:\Users\user\Desktop\log4js-node\test.js`
> `debug-path.sep: \`
> `debug-trunc-filename: test.js`
> `2022-01-04T21:52:38.144 test.js Test log message C:\Users\user\Desktop\log4js-node\test.js`
---

I noticed a difference between yours and mine:
`file:///C:/Users/Username/Desktop/Projects/Logtest/src/main.js` vs `C:\Users\user\Desktop\log4js-node\test.js`

Yours is having the prefix `file:///` and using `/` instead of `\`.
Are you running Windows Subsystem for Linux (WSL) or something?

> ### path.sep
> Added in: v0.7.9
> [<string>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type)
> Provides the platform-specific path segment separator:
> 
> - `\` on Windows
> - `/` on POSIX
>
> For example, on POSIX:
> 
> ```javascript
> 'foo/bar/baz'.split(path.sep);
> // Returns: ['foo', 'bar', 'baz']
> ```
> On Windows:
> 
> ```javascript
> 'foo\\bar\\baz'.split(path.sep);
> // Returns: ['foo', 'bar', 'baz']
> ```
> On Windows, both the forward slash (`/`) and backward slash (`\`) are accepted as path segment separators; however, the `path` methods only add backward slashes (`\`).
>
> (_src: https://nodejs.org/api/path.html#pathsep_)

**Comment by peteriman** [comment]
I ran your codes, with some modifications saved into `main.js`, using `log4js 6.3.0` on `Windows 10` with `Node.js 14`.

```javascript
const log4js = require('./lib/log4js'); // changed

log4js.configure({
  appenders: {
    out: {
      type: 'console',
      layout: {
        type: 'pattern',
        pattern: '%d %f{1} %m',
      },
    },
  },
  categories: {
    default: { appenders: ['out'], level: 'info', enableCallStack: true },
  },
});

const logger = log4js.getLogger('app');

logger.info('Test log message', __filename); // changed
```


## `node main.js`
> v14.15.0 - `2022-01-04T21:45:59.119 main.js Test log message C:\Users\user\Desktop\log4js-node\main.js`
> v14.15.1 - `2022-01-04T21:46:05.225 main.js Test log message C:\Users\user\Desktop\log4js-node\main.js`
> v14.15.2 - `2022-01-04T21:46:08.476 main.js Test log message C:\Users\user\Desktop\log4js-node\main.js`
> v14.15.3 - `2022-01-04T21:46:13.812 main.js Test log message C:\Users\user\Desktop\log4js-node\main.js`
> v14.15.4 - `2022-01-04T21:46:15.480 main.js Test log message C:\Users\user\Desktop\log4js-node\main.js`
> v14.15.5 - `2022-01-04T21:46:18.274 main.js Test log message C:\Users\user\Desktop\log4js-node\main.js`

Doesn't seem to have any issues.

----

I edited the [lib/layouts.js](https://github.com/log4js-node/log4js-node/blob/master/lib/layouts.js) to have some debug logs:
```javascript
  function fileName(loggingEvent, specifier) {
    let filename = loggingEvent.fileName || '';
    console.log("debug-full-filename:", filename);            // added
    console.log("debug-path.sep:", path.sep);                 // added
    if (specifier) {
      const fileDepth = parseInt(specifier, 10);
      const fileList = filename.split(path.sep);
      if (fileList.length > fileDepth) {
        filename = fileList.slice(-fileDepth).join(path.sep);
        console.log("debug-trunc-filename:", filename);       // added
      }
    }

    return filename;
  }
```

## `node main.js`

> `debug-full-filename: C:\Users\user\Desktop\log4js-node\main.js`
> `debug-path.sep: \`
> `debug-trunc-filename: main.js`
> `2022-01-04T21:52:38.144 main.js Test log message C:\Users\user\Desktop\log4js-node\main.js`
---

I noticed a difference between yours and mine:
`file:///C:/Users/Username/Desktop/Projects/Logtest/src/main.js` vs `C:\Users\user\Desktop\log4js-node\main.js`

Yours is having the prefix `file:///` and using `/` instead of `\`.
Are you running Windows Subsystem for Linux (WSL) or something?

> ### path.sep
> Added in: v0.7.9
> [<string>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type)
> Provides the platform-specific path segment separator:
> 
> - `\` on Windows
> - `/` on POSIX
>
> For example, on POSIX:
> 
> ```javascript
> 'foo/bar/baz'.split(path.sep);
> // Returns: ['foo', 'bar', 'baz']
> ```
> On Windows:
> 
> ```javascript
> 'foo\\bar\\baz'.split(path.sep);
> // Returns: ['foo', 'bar', 'baz']
> ```
> On Windows, both the forward slash (`/`) and backward slash (`\`) are accepted as path segment separators; however, the `path` methods only add backward slashes (`\`).
>
> (_src: https://nodejs.org/api/path.html#pathsep_)

**Comment by peteriman** [comment]
@CarlosGomez-dev Can you verify?

**Comment by qyonline** [comment]
@peteriman 
I also met this issue.
May be node stacks causes this issue. (Windows 10, Node v16.14.0)
code:
`console.log(new Error())`
output on windows:
`Error
    at [file:///d:/node/node-practice/app.mjs:6:13]()
    at ModuleJob.run (node:internal/modules/esm/module_job:195:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:337:24)
    at async loadESM (node:internal/process/esm_loader:88:5)`

**Comment by peteriman** [comment]
@qyonline Your error doesn't seem related. Could you provide more of the stack trace and your code?

**Comment by qyonline** [comment]
@peteriman There is no problem in Common JS.
In ES ES module
`import log4js from 'log4js';
...
logger.info('Test log message'
`
output:
![image](https://user-images.githubusercontent.com/16875067/160542669-8cc777aa-9c9a-4c59-8089-c24497a7e3a4.png)

**Comment by peteriman** [comment]
Oh I see, this is because ESM uses `url` instead of `path`.
I'll be patching it. Thanks!

**Comment by peteriman** [comment]
Released in `log4js@6.4.5`.

@CarlosGomez-dev @qyonline Please let me know if there are still issues after upgrading.
